**마크다운**
* '#' : CHAPTER
* '##' : 소제목 (01.~)
* '###' 볼드체
* '####' 갈색 글씨

## 04. 데이터 핸들링 - 판다스

판다스는 파이썬에서 데이터 처리를 위해 존재하는 가장 인기 있는 라이브러리이다. 일반적으로 대부분의 데이터 세트는 2차원 데이터이다. 즉, 행(Row)x열(column)로 구성돼 있다(RDBMS의 TABLE이나 엑셀 시트를 떠올리면 된다). 행과 열의 2차원 데이터가 인기 있는 이유는 바로 인간이 가장 이해하기 쉬운 데이터 구조이면서도 효과적으로 데이터를 담을 수 있는 구조이기 때문이다. 판다스는 이처럼 행과 열로 이루어진 2차원 데이터를 효율적으로 가공/처리할 수 있는 다양하고 훌륭한 기능을 제공한다.

앞 절에서 넘파이를 소개했지만, 넘파이의 데이터 핸들링은 편하다고 말하기 어렵다. 넘파이는 저수준 API가 대부분이다. 판다스는 많은 부분이 넘파이 기반으로 작성됐는데, 넘파이보다 훨씬 유연하고 편리하게 데이터 핸들링을 가능하게 해준다. RDBMS의 SQL이나 엑셀 시트의 편의성만큼은 아니더라도, 판다스는 이에 버금가는 고수준 API를 제공한다(판다스 전문가 중에는 판다스가 RDBMS SQL보다 더 편리하다고 생각하는 사람도 있다). 판다스는 파이썬의 리스트, 컬렉션, 넘파이 등의 내부 데이터뿐만 아니라 CSV 등의 파일을 쉽게 DataFrame으로 변경해 데이터의 가공/분석을 편리하게 수행할 수 있게 만들어준다.

판다스의 핵심 객체는 DataFrame이다. DataFrame은 여러 개의 행과 열로 이뤄진 2차원 데이터를 담는 데이터 구조체이다. 판다스가 다루는 대부분의 영역은 바로 DataFrame에 관련된 부분이다. DataFrame을 이해하기 전에 다른 중요 객체인 Index와 Series를 이해하는 것도 중요하다. Index는 RDBMS의 PK처럼 개별 데이터를 고유하게 식별하는 Key 값이다. Series와 DataFrame은 모두 Index를 key 값으로 가지고 있다. Series와 DataFrame의 가장 큰 차이는 Series는 칼럼이 하나뿐인 데이터 구조체이고, DataFrame은 칼럼이 여러 개인 데이터 구조체라는 점이다.

### 판다스 시작 - 파일을 DataFrame으로 로딩, 기본 API

pandas를 pd로 에일리어스(alias)해 임포트하는 것이 관례이다.

In [ ]:
import pandas as pd

판다스는 다양한 포맷으로 된 파일을 DataFrame으로 로딩할 수 있는 편리한 API를 제공한다. 대표적으로 read_csv(), read_table(), read_fwf()가 있다.
* read_csv() : 이름에서도 알 수 있듯이 CSV(칼럼을 ','으로 구분한 파일 포맷) 파일 포맷 변환을 위한 API이다.

read_table()과 read_csv()의 가장 큰 차이는 필드 구분 문자(Delimeter)가 콤마(',')냐, 탭('\t')이냐의 차이다.
* read_table()의 디폴트 필드 구분 문자는 탭 문자이다.

read_csv()는 CSV뿐만 아니라 어떤 필드 구분 문자 기반의 파일 포맷도 DataFrame으로 변환이 가능하다. read_csv()의 인자인 sep에 해당 구분 문자를 입력하면 된다. 가령 탭으로 필드가 구분돼 있다면 read_csv('파일명', sep='\t')처럼 쓰면 된다. read_csv()에서 sep 인자를 생략하면 자동으로 콤마로 할당한다(sep=','). read_table()은 기능상 큰 차이가 없다.

* read_fwf()는 Fixed Width, 즉 고정 길이 기반의 칼럼 포맷을 DataFrame으로 로딩하기 위한 API이다.

read_csv(filepath_or_buffer, sep=',', ...) 함수에서 가장 중요한 인자는 filepath이다. 나머지 인자는 지정하지 않으면 디폴트 값으로 할당된다. Filepath에는 로드하려는 데이터 파일의 경로를 포함한 파일명을 입력하면 된다. 파일명을 입력되면 파이썬 실행 파일이 있는 디렉터리와 동일한 디렉터리에 있는 파일명을 로딩한다.

In [ ]:
titanic_df=pd.read_csv('/content/sample_data/titanic_train.csv') ## 노트북에 저장하면 오류가 떠서 colab 파일에 저장한 후 불러옴.
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

In [ ]:
print('titanic 변수 type', type(titanic_df))
titanic_df

titanic 변수 type <class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass  \
0              1         0       3   
1              2         1       1   
2              3         1       3   
3              4         1       1   
4              5         0       3   
..           ...       ...     ...   
886          887         0       2   
887          888         1       1   
888          889         0       3   
889          890         1       1   
890          891         0       3   

                                                  Name     Sex   Age  SibSp  \
0                              Braund, Mr. Owen Harris    male  22.0      1   
1    Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                               Heikkinen, Miss. Laina  female  26.0      0   
3         Futrelle, Mrs. Jacques Heath (Lily May Peel)  female  35.0      1   
4                             Allen, Mr. William Henry    male  35.0      0   
..                                                 ...     ...   ...    ...   
886                              Montvila, Rev. Juozas    male  27.0      0   
887                       Graham, Miss. Margaret Edith  female  19.0      0   
888           Johnston, Miss. Catherine Helen "Carrie"  female   NaN      1   
889                              Behr, Mr. Karl Howell    male  26.0      0   
890                                Dooley, Mr. Patrick    male  32.0      0   

     Parch            Ticket     Fare Cabin Embarked  
0        0         A/5 21171   7.2500   NaN        S  
1        0          PC 17599  71.2833   C85        C  
2        0  STON/O2. 3101282   7.9250   NaN        S  
3        0            113803  53.1000  C123        S  
4        0            373450   8.0500   NaN        S  
..     ...               ...      ...   ...      ...  
886      0            211536  13.0000   NaN        S  
887      0            112053  30.0000   B42        S  
888      2        W./C. 6607  23.4500   NaN        S  
889      0            111369  30.0000  C148        C  
890      0            370376   7.7500   NaN        Q  

[891 rows x 12 columns]

pd.read_csv()는 호출 시 파일명 인자로 들어온 파일을 로딩해 DataFrame 객체로 반환한다. DataFrame 객체를 살펴보면 데이터 파일의 첫 번째 줄에 있던 칼럼 문자열이 DataFrame의 칼럼으로 할당되었다. read_csv()는 별다른 파라미터 지정이 없으면 파일의 맨 처음 로우를 칼럼명으로 인지하고 칼럼으로 변환한다. 그리고 콤마로 분리된 데이터들이 해당 칼럼에 맞게 할당되었다. 그런데 맨 왼쪽을 보면 파일에 기재돼 있지 않는 데이터값이 로우 순으로 0, 1, 2, 3, ...과 같이 순차적으로 표시되어 있음을 알 수 있다(맨 처음 칼럼인 Passenger2) 칼럼의 왼쪽). 칼럼명도 표시되지 않는 이 데이터들은 바로 판다스의 Index 객체 값이다. 모든 DataFrame 내의 데이터는 생성되는 순간 고유의 Index 값을 가지게 된다. 인덱스는 RDBMS의 PK와 유사하게 고유의 레코드를 식별하는 역할을 한다.

DataFrame의 모든 데이터를 주피터 노트북에 표출하면 보기에 불편할 뿐만 아니라 메모리도 많이 사용하므로 앞으로는 일부 데이터만 표출하도록 하겠다. DataFrame.head()는 DataFrame의 맨 앞에 있는 N개의 로우를 반환한다. head(3)은 맨 앞 3개의 로우를 반환한다(Default는 5개).

In [ ]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

DataFrame의 행과 열의 크기를 알아보는 가장 좋은 방법은 생성된 DataFrame 객체의 shape 변수를 이용하는 것이다. shape은 DataFrame의 행과 열을 튜플 형태로 반환한다.

In [ ]:
print('DataFrame 크기:', titanic_df.shape)

DataFrame 크기: (891, 12)


생성된 DataFrame 객체인 titanic_df는 891개의 로우와 12개의 칼럼으로 이루어져있다.

DataFrame은 데이터뿐만 아니라 칼럼의 타입, Null 데이터 개수, 데이터 분포도 등의 메타 데이터 등도 조회가 가능하다. 이를 위한 대표적인 메서드로 info()와 describe()가 있다.

* info()

In [ ]:
titanic_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 891 entries, 0 to 890
Data columns (total 12 columns):
 #   Column       Non-Null Count  Dtype  
---  ------       --------------  -----  
 0   PassengerId  891 non-null    int64  
 1   Survived     891 non-null    int64  
 2   Pclass       891 non-null    int64  
 3   Name         891 non-null    object 
 4   Sex          891 non-null    object 
 5   Age          714 non-null    float64
 6   SibSp        891 non-null    int64  
 7   Parch        891 non-null    int64  
 8   Ticket       891 non-null    object 
 9   Fare         891 non-null    float64
 10  Cabin        204 non-null    object 
 11  Embarked     889 non-null    object 
dtypes: float64(2), int64(5), object(5)
memory usage: 83.7+ KB


1. RangeIndex : DataFrame index의 범위를 나타내므로 전체 row 수를 알 수 있다. 전체 데이터는 891개 row이다.
2. Data columns : Column 수는 12개이다.
3. Dtype : 칼럼별 데이터 타입. 가령 Pclass 칼럼의 데이터 타입은 int64형이다. Name 칼럼의 데이터 타입은 object인데, 이는 문자열 타입으로 생각해도 무방하다.
4. Non-Null Count : 몇 개의 데이터가 non-null(Null 값이 아님)인지 나타낸다. 가령 Age 칼럼의 714 non-null의 의미는 Age 칼럼 891개 데이터 중 714개가 Null이 아니며 177개는 Null이라는 의미이다.
5. dtypes : 전체 12개의 칼럼들의 타입을 요약한 것이다. 2개의 칼럼이 float64, 5개의 칼럼이 int64, 5개의 칼럼이 object 타입이다.

info() 메서드를 통해서 총 데이터 건수와 데이터 타입, Null 건수를 알 수 있다.

* describe() : describe() 메서드는 칼럼별 숫자형 데이터값의 n-percentile 분포도, 평균값, 최댓값, 최솟값을 나타낸다. describe() 메서드는 오직 숫자형(int, float 등) 칼럼의 분포도만 조사하며 자동으로 object 타입의 칼럼은 출력에서 제외시킨다.

데이터의 분포도를 아는 것은 머신러닝 알고리즘 능력을 향상시키는 중요한 요소이다. 가령 회귀에서 결정 값이 정규 분포를 이루지 않고 특정 값으로 왜곡돼 있는 경우, 또는 데이터값에 이상치가 많을 경우 예측 성능이 저하된다. describe() 메서드만으로 정확한 분포도를 알기는 무리지만, 개략적인 수준의 분포도를 알 수 있어 유용하다.

In [ ]:
titanic_df.describe()

PassengerId    Survived      Pclass         Age       SibSp  \
count   891.000000  891.000000  891.000000  714.000000  891.000000   
mean    446.000000    0.383838    2.308642   29.699118    0.523008   
std     257.353842    0.486592    0.836071   14.526497    1.102743   
min       1.000000    0.000000    1.000000    0.420000    0.000000   
25%     223.500000    0.000000    2.000000   20.125000    0.000000   
50%     446.000000    0.000000    3.000000   28.000000    0.000000   
75%     668.500000    1.000000    3.000000   38.000000    1.000000   
max     891.000000    1.000000    3.000000   80.000000    8.000000   

            Parch        Fare  
count  891.000000  891.000000  
mean     0.381594   32.204208  
std      0.806057   49.693429  
min      0.000000    0.000000  
25%      0.000000    7.910400  
50%      0.000000   14.454200  
75%      0.000000   31.000000  
max      6.000000  512.329200

* count : Not Null인 데이터 건수
* mean : 전체 데이터의 평균값
* std : 표준편차
* min : 최솟값
* max : 최댓값
* 25% : 25 percentile 값
* 50% : 50 percentile 값
* 75% : 75 percentile 값

describe() 해당 숫자 칼럼이 숫자형 카테고리 칼럼인지를 판단할 수 있게 도와준다. 카테고리 칼럼은 특정 범주에 속하는 값을 코드화한 칼럼이다. 가량 성별 칼럼의 경우 '남'과 '여'가 있고 '남'을 1, '여'를 2와 같이 표현한 것이다. 이러한 카테고리 칼럼을 숫자로 표시할 수도 있다.

* PassengerID : 승객 ID를 식별하는 칼럼이므로 1~891까지 숫자가 할당되어서 분석을 위한 의미 있는 속성이 아니다.
* Survived : min 0, 25%~75%가 2, 3, max가 3이므로 1, 2, 3으로 이뤄진 숫자형 카테고리 칼럼일 것이다.
* Pclass : DataFrame의 [] 연산자 내부에 칼럼명을 입력하면 Series 형태로 특정 칼럼 데이터 세트가 반환된다. 이렇게 반환된 **Series 객체에 `value_counts()` 메서드를 호출하면 해당 칼럼값의 유형과 건수를 확인할 수 있다. `value_counts()`는 지정된 칼럼의 데이터값 건수를 반환한다.** `value_counts()`는 데이터의 분포도를 확인하는 데 매우 유용한 함수이다.

In [ ]:
value_counts=titanic_df['Pclass'].value_counts()
print(value_counts)

3    491
1    216
2    184
Name: Pclass, dtype: int64


value_counts()의 반환 결과는 Pclass값 3이 491개, 2가 184개이다. value_counts()는 많은 건수 순서로 정렬되어 값을 반환한다. DataFrame의 [] 연산자 내부에 칼럼명을 입력하면 해당 칼럼에 해당하는 Series 객체를 반환한다.

In [ ]:
titanic_pclass=titanic_df['Pclass']
print(type(titanic_pclass))

<class 'pandas.core.series.Series'>


Series는 Index와 단 하나의 칼럼으로 구성된 데이터 세트이다. 이렇게 반환된 Series 객체가 어떤 값으로 구성되어 있는지 알아보자. 전체 891개 데이터를 추출하지 않고 head() 메서드를 이용해 앞의 5개만 추출해보자.

In [ ]:
titanic_pclass.head()

0    3
1    1
2    3
3    1
4    3
Name: Pclass, dtype: int64

* 왼쪽 열 : Series의 Index
* 오른쪽 열 : Series의 데이터 값

한 개 칼럼의 데이터만 출력되지 않고 맨 왼쪽에 0부터 시작하는 순차 값이 있다. 이것은 DataFrame의 인덱스와 동일한 인덱스이다. 오른쪽은 Series의 해당 칼럼의 데이터값이다. 모든 Series와 DataFrame은 인덱스를 반드시 가진다. 방금 소개한 value_counts() 메서드는 Series 객체에서만 정의돼 있다. DataFrame은 value_counts() 메서드를 가지고 있지 않다.

* value_counts() 메서드의 데이터 타입과 반환 값을 다시 확인해보자.

In [ ]:
value_counts=titanic_df['Pclass'].value_counts()
print(type(value_counts))
print(value_counts)

<class 'pandas.core.series.Series'>
3    491
1    216
2    184
Name: Pclass, dtype: int64


value_counts()가 반환하는 데이터 타입 역시 Series 객체이다. 이 반환된 Series 객체의 값을 보면 맨 왼쪽이 인덱스값이며, 오른쪽이 데이터값이다. 그런데 이번에는 인덱스가 단순히 0부터 시작하는 순차 값이 아니다. Pclass 칼럼 값이 3, 1, 2를 나타내고 있다. 이처럼 **인덱스는 단순히 순차 값과 같은 의미 없는 식별자만 할당하는 것이 아니라 고유성이 보장된다면 의미 있는 데이터값 할당도 가능하다.** value_counts()는 칼럼 값별 데이터 건수를 반환하므로 고유 칼럼 값을 식별자로 사용할 수 있다. 인덱스는 DataFrame, Series가 만들어진 후에도 변경할 수 있다. 인덱스는 또한 숫자형뿐만 아니라 문자열도 가능하다. 단, 모든 인덱스는 고유성이 보장되어야 한다.

### DataFrame과 리스트, 딕셔너리, 넘파이 ndarray 상호 변환

기본적으로 DataFrame은 파이썬의 리스트, 딕셔너리 그리고 넘파이 ndarray 등 다양한 데이터로부터 생성될 수 있다. 또한 DataFrame은 반대로 파이썬의 리스트, 딕셔너리 그리고 넘파이 ndarray 등으로 변환될 수 있다. 특히 사이킷런의 많은 API는 DataFrame을 인자로 입력 받을 수 있지만, 기본적으로 넘파이 ndarray를 입력 인자로 사용하는 경우가 대부분이다. 따라서 DataFrame과 넘파이 ndarray 상호 간의 변환은 매우 빈번하게 발생한다.

api 설명 참고 사이트 : https://blog.wishket.com/api%EB%9E%80-%EC%89%BD%EA%B2%8C-%EC%84%A4%EB%AA%85-%EA%B7%B8%EB%A6%B0%ED%81%B4%EB%9D%BC%EC%9D%B4%EC%96%B8%ED%8A%B8/

#### 넘파이 ndarray, 리스트, 딕셔너리를 DataFrame으로 변환하기

넘파이 ndarray, 파이썬 리스트, 딕셔너리로부터 DataFrame을 생성해 보자. DataFrame은 리스트와 넘파이 ndarray와 다르게 칼럼명을 가지고 있다. 이 칼럼명으로 인하여 리스트와 넘파이 ndarray보다 상대적으로 편하게 데이터 핸들링이 가능하다. 일반적으로 DataFrame으로 변환 시 이 칼럼명을 지정해 준다(지정하지 않으면 자동으로 칼럼명을 할당한다). 판다스 DataFrame 객체의 생성 인자 data는 리스트나 딕셔너리 또는 넘파이 ndarray를 입력 받고, 생성 인자 columns는 칼럼명 리스트를 입력 받아서 쉽게 DataFrame을 생성할 수 있다.

DataFrame은 기본적으로 행과 열을 가지는 2차원 데이터다. 따라서 2차원 이하의 데이터들만 DataFrame으로 변환될 수 있다.

* 먼저 1차원 형태의 리스트와 넘파이 ndarray부터 DataFrame으로 변환해 보자. 1차원 데이터이므로 칼럼은 1개만 필요하며, 칼럼명을 'col1'으로 지정한다.

In [ ]:
import numpy as np

col_name1=['col1']
list1=[1, 2, 3]
array1=np.array(list1)
print('array1 shape:', array1.shape)

# 리스트를 이용해 DataFrame 생성
df_list1=pd.DataFrame(list1, columns=col_name1)
print('1차원 리스트로 만든 DataFrame:\n', df_list1)

# 넘파이 ndarray를 이용해 DataFrame 생성
df_array1=pd.DataFrame(array1, columns=col_name1)
print('1차원 ndarray로 만든 DataFrame:\n', df_array1)

array1 shape: (3,)
1차원 리스트로 만든 DataFrame:
    col1
0     1
1     2
2     3
1차원 ndarray로 만든 DataFrame:
    col1
0     1
1     2
2     3


1차원 형태의 데이터를 기반으로 DataFrame을 생성하므로 칼럼명이 한 개만 필요하다는 사실에 주의해야 한다.

* 2차원 형태의 데이터를 기반으로 DataFrame을 생성해 보자. 2행 3열 형태의 리스트와 ndarray를 기반으로 DataFrame을 생성하므로 칼럼명은 3개가 필요하다.

In [ ]:
# 3개의 칼럼명이 필요함
col_name2=['col1', 'col2', 'col3']

# 2행x3열 형태의 리스트와 ndarray 생성한 뒤 이를 DataFrame으로 변환
list2=[[1, 2, 3],
       [11, 12, 13]]
array2=np.array(list2)
print('array2 shape:', array2.shape)
df_list2=pd.DataFrame(list2, columns=col_name2)
print('2차원 리스트로 만든 DataFrame:\n', df_list2)
df_array2=pd.DataFrame(array2, columns=col_name2)
print('2차원 ndarray로 만든 DataFrame:\n', df_array2)

array2 shape: (2, 3)
2차원 리스트로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13
2차원 ndarray로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    12    13


* 딕셔너리를 DataFrame으로 변환해 보자. 일반적으로 딕셔너리를 DataFrame으로 변환 시에는 딕셔너리의 키(Key)는 칼럼명으로, 딕셔너리의 값(Value)는 키에 해당하는 칼럼 데이터로 변환된다. 따라서 키의 경우는 문자열, 값의 경우 리스트(또는 ndarray) 형태로 덱셔너리를 구성한다.

In [ ]:
# Key는 문자열 칼럼명으로 매핑, Value는 리스트 형(또는 ndarray) 칼럼 데이터로 매핑
dict={'col1':[1, 11], 'col2':[2, 22], 'col3':[3, 33]}
df_dict=pd.DataFrame(dict)
print('딕셔너리로 만든 DataFrame:\n', df_dict)

딕셔너리로 만든 DataFrame:
    col1  col2  col3
0     1     2     3
1    11    22    33


Key 값은 칼럼명, 각 Value는 각 칼럼 데이터로 매핑되었음을 알 수 있다.

#### DataFrame을 넘파이 ndarray, 리스트, 딕셔너리로 변환하기

많은 머신러닝 패키지가 기본 데이터 형으로 넘파이 ndarray를 사용한다. 따라서 데이터 핸들링은 DataFrame을 이용하더라도 머신러닝 패키지의 입력 인자 등에 적용하기 위해 다시 넘파이 ndarray로 변환하는 경우가 빈번하게 발생한다. **DataFrame을 넘파이 ndarray로 변환하는 것은 DataFrame 객체의 values를 이용해 쉽게 할 수 있다.** 이 책 전반에 거쳐서 values를 이용한 ndarray로의 변환은 매우 많이 사용된다.
* 방금 전에 생성한 DataFrame객체인 df_dict를 ndarray로 변환해 보자.

In [ ]:
# DataFrame을 ndarray로 변환
array3=df_dict.values
print('df_dict.values 타입:', type(array3), 'df_dict.values shape:', array3.shape)
print(array3)

df_dict.values 타입: <class 'numpy.ndarray'> df_dict.values shape: (2, 3)
[[ 1  2  3]
 [11 22 33]]


* DataFrame을 리스트와 딕셔너리로 반환해 보자. 리스트로의 변환은 values로 얻은 ndarray에 `tolist()`를 호출하면 된다. 딕셔너리로의 변환은 DataFrame 객체의 to_dict() 메서드를 호출하는데, 인자로 'list'를 입력하면 딕셔너리의 값이 리스트형으로 반환된다.

In [ ]:
# DataFrame을 리스트로 변환
list3=df_dict.values.tolist()
print('df_dict.values.tolist() 타입:', type(list3))
print(list3)

# DataFrame을 딕셔너리로 변환
dict3=df_dict.to_dict('list')
print('\n df_dict.to_dict() 타입:', type(dict3))
print(dict3)

df_dict.values.tolist() 타입: <class 'list'>
[[1, 2, 3], [11, 22, 33]]

 df_dict.to_dict() 타입: <class 'dict'>
{'col1': [1, 11], 'col2': [2, 22], 'col3': [3, 33]}


### DataFrame의 칼럼 데이터 세트 생성과 수정

DataFrame의 칼럼 데이터 세트 생성과 수정 역시 [] 연산자를 이용해 쉽게 할 수 있다.
* 먼저 Titanic DataFrame의 새로운 칼럼 Age_0을 추가하고 일괄적으로 0 값을 할당하겠다. DataFrame [] 내에 새로운 칼럼명을 입력하고 값을 할당해주기만 하면 된다.

In [ ]:
titanic_df['Age_0']=0
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  
0      0         A/5 21171   7.2500   NaN        S      0  
1      0          PC 17599  71.2833   C85        C      0  
2      0  STON/O2. 3101282   7.9250   NaN        S      0

새로운 칼럼명 'Age_0'으로 모든 데이터값이 0으로 할당된 Series가 기존 DataFrame에 추가됨을 알 수 있다. 이렇듯 칼럼 Series에 값을 할당하고 DataFrame에 추가하는 것은 판다스에서 매우 간단하다. titanic_df['Age_0']=0과 같이 Series에 상숫값을 할당하면 Series의 모든 데이터 세트에 일괄적으로 적용된다. 이는 넘파이의 ndarray에 상숫값을 할당할 때 모든 ndarray 값에 일괄 적용됨과 동일하다.
* 기존 칼럼 Series의 데이터를 이용해 새로운 칼럼 Series를 만들겠다.

In [ ]:
titanic_df['Age_by_10']=titanic_df['Age']*10
titanic_df['Family_No']=titanic_df['SibSp']+titanic_df['Parch']+1
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      220.0   
1      0          PC 17599  71.2833   C85        C      0      380.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      260.0   

   Family_No  
0          2  
1          2  
2          1

기존 칼럼 Series를 가공해 새로운 칼럼 Series인 Age_by_10과 Family_No가 새롭게 DataFrame에 추가됨을 알 수 있다.

DataFrame 내의 기존 칼럼 값도 쉽게 일괄적으로 업데이트할 수 있다. 업데이트를 원하는 칼럼 Series를 DataFrame [] 내에 칼럼 명으로 입력한 뒤에 값을 할당해주면 된다.
* 새롭게 추가한 'Age_by_10' 칼럼 값을 일괄적으로 기존 값+100으로 업데이트해 보겠다.

In [ ]:
titanic_df['Age_by_10']=titanic_df['Age_by_10']+100
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

Age_by_10 칼럼의 모든 값이 기존 값+100의 값으로 업데이트됨을 알 수 있다.

### DataFrame 데이터 삭제

DataFrame에서 데이터의 삭제는 drop() 메서드를 이용한다.drop() 메서드는 사용하는 데 혼동을 줄 만한 부분이 있어서 주의가 필요하다.

`DataFrame.drop(labels=None, axis=0, index=None, columns=None, inplace=False, errors='raise')`

가장 중요한 파라미터는 labels, axis, inplace이다. 먼저 axis 값에 따라서 특정 칼럼 또는 특정 행을 드롭한다. **axis 0은 로우 방향 축, axis 1은 칼럼 방향 축**이다. 판다스의 DataFrame은 2차원 데이터만 다루므로 axis 0, axis 1로만 axis가 구성돼 있다.

따라서 drop() 메서드에 axis=1을 입력하면 칼럼 축 방향으로 드롭을 수행하므로 칼럼을 드롭하겠다는 의미이다. labels에 원하는 칼럼 명을 입력하고 axis=1을 입력하면 지정된 칼럼을 드롭한다. drop() 메서드에 axis=0을 입력하면 로우 축 방향으로 드롭을 수행하므로 특정 로우를 드롭하겠다는 것이다. DataFrame의 특정 로우를 가리키는 것은 인덱스이다. 따라서 axis를 0으로 지정하면 DataFrame은 자동으로 labels에 오는 값을 인덱스로 간주한다. 이 책에서 drop() 메서드가 사용되는 대부분의 경우는 칼럼을 드롭하는 경우이다. 기존 칼럼 값을 가공해 새로운 칼럼을 만들고 삭제하는 경우가 많다 보니 axis=1로 설정하고 드롭하는 경우가 많을 수밖에 없다. axis=0으로 설정하고 로우 레벨로 삭제를 하는 경우는 이상치 데이터를 삭제하는 경우에 주로 사용된다.

* 지난 예제에서 새롭게 Titanic DataFrame에 추가된 'Age_0' 칼럼을 삭제해 보자.

In [ ]:
titanic_drop_df=titanic_df.drop('Age_0', axis=1)
titanic_drop_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_by_10  Family_No  
0      0         A/5 21171   7.2500   NaN        S      320.0          2  
1      0          PC 17599  71.2833   C85        C      480.0          2  
2      0  STON/O2. 3101282   7.9250   NaN        S      360.0          1

* 주요 파라미터인 inplace에 대해 알아보기 위해 원본 Titanic DataFrame을 다시 확인해 보자.

In [ ]:
titanic_df.head(3)

PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  Age_0  Age_by_10  \
0      0         A/5 21171   7.2500   NaN        S      0      320.0   
1      0          PC 17599  71.2833   C85        C      0      480.0   
2      0  STON/O2. 3101282   7.9250   NaN        S      0      360.0   

   Family_No  
0          2  
1          2  
2          1

삭제됐다고 생각했던 'Age_0' 칼럼이 여전히 존재한다. 그 이유는 앞의 예제 코드에서 inplce=Fale로 설정했기 때문이다(inplace는 디폴트 값이 False이므로 inplace 파라미터를 기재하지 않으면 자동으로 False가 된다). Inplace=False이면 자기 자신의 DataFrame의 데이터는 삭제하지 않으며, 삭제된 결과 DataFrame을 반환한다. 그래서 titanic_df DataFrame 객체 변수는 'Age_0' 칼럼이 삭제되지 않았으며 titanic_df.drop('Age_0', axis=1)의 수행 결과로 반환된 titanic_drop_df DataFrame 객체 변수만 'Age_0' 칼럼이 삭제된 것이다.

inplace=True로 설정하면 자신의 DataFrame의 데이터를 삭제한다. 또한 여러 개의 칼럼을 삭제하고 싶으면 리스트 형태로 삭제하고자 하는 칼럼 명을 입력해 labels 파라미터로 입력하면 된다.
* 이번에는 titanic_df의 'Age_0', 'Age_by_10', 'Family_No' 칼럼을 모두 삭제하겠다.

In [ ]:
drop_result=titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis=1, inplace=True)
print('inplace=True로 drop 후 반환된 값:', drop_result)
titanic_df.head(3)

inplace=True로 drop 후 반환된 값: None


PassengerId  Survived  Pclass  \
0            1         0       3   
1            2         1       1   
2            3         1       3   

                                                Name     Sex   Age  SibSp  \
0                            Braund, Mr. Owen Harris    male  22.0      1   
1  Cumings, Mrs. John Bradley (Florence Briggs Th...  female  38.0      1   
2                             Heikkinen, Miss. Laina  female  26.0      0   

   Parch            Ticket     Fare Cabin Embarked  
0      0         A/5 21171   7.2500   NaN        S  
1      0          PC 17599  71.2833   C85        C  
2      0  STON/O2. 3101282   7.9250   NaN        S

한 가지 유의할 점은 drop() 시 inplace=True로 설정하면 반환 값이 None(아무 값도 아님)이 된다. 따라서 다음 코드와 같이 inplace=True로 설정한 채로 반환 값을 다시 자신의 DataFrame 객체로 할당하면 안 된다. titanic_df=titanic_df.drop(['Age_0', 'Age_by_10', 'Family_No'], axis=1, inplace=True)는 titanic_df 객체 변수를 아예 None으로 만들어 버린다.

* axis=0으로 설정해 index 0, 1, 2(맨 앞 3개 데이터) 로우를 삭제해 보자.

In [ ]:
pd.set_option('display.width', 1000)
pd.set_option('display.max_colwidth', 15)
print('#### before axis 0 drop ####')
print(titanic_df.head(3))

titanic_df.drop([0, 1, 2], axis=0, inplace=True)

print('#### after axis 0 drop ####')
print(titanic_df.head(3))

#### before axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S
#### after axis 0 drop ####
   PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
3            4         1       1  Futrelle, M...  female  35.0      1      0  113803  53.1000  C123        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0  373450   8.0500   NaN        S
5            6         0       3  Moran, Mr. ...    male   NaN      0      0  330877   8.4583   NaN        Q


Index 0, 1, 2에 위치한 로우를 titanic_df.drop([0, 1, 2], axis=0, inplace=True)로 삭제했다. 삭제 후에 head(3)으로 확인한 맨 앞 3개 데이터의 인덱스는 3, 4, 5이다. 인덱스 0, 1, 2에 위치한 로우가 삭제됐음을 알 수 있다. 다음과 같이 drop() 메서드에서 axis와 inplace 인자를 적용해 DataFrame을 변경하는 방식을 정리할 수 있다.

* axis : DataFrame의 로우를 삭제할 때는 axis=0, 칼럼을 삭제할 때는 axis=1로 설정
* 원본 DataFrame은 유지하고 드롭된 DataFrame을 새롭게 객체 변수로 받고 싶다면 inplace=False로 설정(디폴트 값이 False).

예 : titanic_drop_df=titanic_df.drop('Age_0', axis=1, inplace=False)
* 원본 DataFrame에 드롭된 결과를 적용할 경우 inplace=True를 적용

예 : titanic_df.drop('Age_0', axis=1, inplace=True)
* 원본 DataFrame에서 드롭된 DataFrame을 다시 원본 DataFrame 객체 변수로 할당하면 원본 DataFrame에서 드롭된 결과를 적용할 경우와 같음(단, 기존 원본 DataFrame 객체 변수는 메모리에서 추후 제거됨).

예 : titanic_df=titanic_df.drop('Age_0', axis=1, inplace=False)

### Index 객체
판다스의 Index 객체는 RDBMS의 PK(Primary Key)와 유사하게 DataFrame, Series의 레코드를 고유하게 식별하는 객체이다. DataFrame, Series에서 Index 객체만 추출하려면 DataFrame.index 또는 Series.index 속성을 통해 가능하다.
* titanic_df DataFrame에서 Index 객체를 추출해 보자.

drop으로 원본 Titanic DataFrame의 일부 데이터가 삭제됐으므로 다시 csv 파일에서 로딩한 후 수행하겠다. 반환된 Index 객체의 실제 값은 넘파이 1차원 ndarray로 볼 수 있다. Index 객체의 values 속성으로 ndarray 값을 알 수 있다.

In [ ]:
# 원본 파일 다시 로딩
titanic_df=pd.read_csv('/content/sample_data/titanic_train.csv')

# Index 객체 추출
indexes=titanic_df.index
print(indexes)

# Index 객체를 실제 값 array로 변환
print('Index 객체 array값:\n', indexes.values)

RangeIndex(start=0, stop=891, step=1)
Index 객체 array값:
 [  0   1   2   3   4   5   6   7   8   9  10  11  12  13  14  15  16  17
  18  19  20  21  22  23  24  25  26  27  28  29  30  31  32  33  34  35
  36  37  38  39  40  41  42  43  44  45  46  47  48  49  50  51  52  53
  54  55  56  57  58  59  60  61  62  63  64  65  66  67  68  69  70  71
  72  73  74  75  76  77  78  79  80  81  82  83  84  85  86  87  88  89
  90  91  92  93  94  95  96  97  98  99 100 101 102 103 104 105 106 107
 108 109 110 111 112 113 114 115 116 117 118 119 120 121 122 123 124 125
 126 127 128 129 130 131 132 133 134 135 136 137 138 139 140 141 142 143
 144 145 146 147 148 149 150 151 152 153 154 155 156 157 158 159 160 161
 162 163 164 165 166 167 168 169 170 171 172 173 174 175 176 177 178 179
 180 181 182 183 184 185 186 187 188 189 190 191 192 193 194 195 196 197
 198 199 200 201 202 203 204 205 206 207 208 209 210 211 212 213 214 215
 216 217 218 219 220 221 222 223 224 225 226 227 228 229 230 231 232

Index 객체는 식별성 데이터를 1차원 array로 가지고 있다. 또한 ndarray와 유사하게 단일 값 반환 및 슬라이싱도 가능하다.

In [ ]:
print(type(indexes.values))
print(indexes.values.shape)
print(indexes[:5].values)
print(indexes.values[:5])
print(indexes[6])

<class 'numpy.ndarray'>
(891,)
[0 1 2 3 4]
[0 1 2 3 4]
6


한 번 만들어진 DataFrame 및 Series의 Index 객체는 함부로 변경할 수 없다. 즉, 다음과 같이 첫 번째 Index 객체의 값을 5로 변경하는 작업은 수행할 수 없다.

In [ ]:
# indexes[0]=5

Series 객체는 Index 객체를 포함하지만 Series 객체에 연산 함수를 적용할 때 Index는 연산에서 제외된다. Index는 오직 식별용으로만 사용된다.

In [ ]:
series_fair=titanic_df['Fare']
print('Fair Series max 값:', series_fair.max())
print('Fair Series sum 값:', series_fair.sum())
print('sum() Fair Series:', sum(series_fair))
print('Fair Series + 3:\n', (series_fair+3).head(3))

Fair Series max 값: 512.3292
Fair Series sum 값: 28693.9493
sum() Fair Series: 28693.949299999967
Fair Series + 3:
 0    10.2500
1    74.2833
2    10.9250
Name: Fare, dtype: float64


DataFrame 및 Series에 reset_index() 메서드를 수행하면 새롭게 인덱스를 연속 숫자 형으로 할당하며 기존 인덱스는 'index'라는 새로운 칼럼 명으로 추가한다.

In [ ]:
titanic_reset_df=titanic_df.reset_index(inplace=False)
titanic_reset_df.head(3)

index  PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0      0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500   NaN        S
1      1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2      2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250   NaN        S

기존 index는 index라는 칼럼으로 추가된다.

reset_index()는 인덱스가 연속된 int 숫자형 데이터가 아닐 경우에 다시 이를 연속 int 숫자형 데이터로 만들 때 주로 사용한다. 예를 들어 Series와 value_counts()를 소개하는 예제에서 'Pclass' 칼럼 Series의 value_counts()를 수행하면 'Pclass' 고유 값이 식별자 인덱스 역할을 했다. 이보다는 연속 숫자형 인덱스가 고유 식별자로 더 적합해 보인다.
* reset_index()를 이용해 인덱스를 다시 만들겠다.

Series에 reset_index()를 적용하면 Series가 아닌 DataFrame이 반환된다. 기존 인덱스가 칼럼으로 추가돼 칼럼이 2개가 되므로 Series가 아닌 DataFrame이 반환된다.

In [ ]:
print('### before reset_index ###')
value_counts=titanic_df['Pclass'].value_counts()
print(value_counts)
print('value_counts 객체 변수 타입:', type(value_counts))
new_value_counts=value_counts.reset_index(inplace=False)
print('### After reset_index ###')
print(new_value_counts)
print('new_value_counts 객체 변수 타입:', type(new_value_counts))

### before reset_index ###
3    491
1    216
2    184
Name: Pclass, dtype: int64
value_counts 객체 변수 타입: <class 'pandas.core.series.Series'>
### After reset_index ###
   index  Pclass
0      3     491
1      1     216
2      2     184
new_value_counts 객체 변수 타입: <class 'pandas.core.frame.DataFrame'>


Series에 reset_index()를 적용하면 새롭게 연속 숫자형 인덱스가 만들어지고 기존 인덱스는 'index' 칼럼명으로 추가되면서 DataFrame으로 변환됨을 알 수 있다. reset_index()의 parameter 중 drop=True로 설정하면 기존 인덱스는 새로운 칼럼으로 추가되지 않고 삭제(drop)된다. 새로운 칼럼이 추가되지 않으므로 그대로 Series로 유지된다.

### 데이터 셀렉션 및 필터링

넘파이의 경우 '[]' 연산자 내 단일 값 추출, 슬라이싱, 팬시 인덱싱, 불린 인덱싱을 통해 데이터를 추출했다 판다스의 경우 ix[], iloc[], loc[] 연산자를 통해 동일한 작업을 수행한다. 판다스의 [] 연산자가 넘파이의 []와 어떤 차이가 있는지 알아보자.

#### DataFrame의 [] 연산자

넘파이에서 [] 연산자는 행의 위치, 열의 위치, 슬라이싱 범위 등을 지정해 데이터를 가져올 수 있었다. 하지만 DataFrame 바로 뒤에 있는 '[]' 안에 들어갈 수 있는 것은 칼럼 명 문자(또는 칼럼 명의 리스트 객체), 또는 인덱스로 변환 가능한 표현식이다. 인덱스로 변환 가능한 표현식이라는 의미는 추후에 다시 설명하겠다. 현재 수준에서는 DataFrame 뒤에 있는 []는 칼럼만 지정할 수 있는 '칼럼 지정 연산자'로 이해하는 게 혼돈을 막는 가장 좋은 방법이다.

titanic_df를 통해 이 차이를 확인해 보겠다. 다음 예제에서 DataFrame에 ['칼럼명']으로 '칼럼명'에 해당하는 칼럼 데이터의 일부만 추출해 보겠다. 여러 개의 칼럼에서 데이터를 추출하려면 ['칼럼1', '칼럼2']와 같이 리스트 객체를 이용하면 된다. 하지만 titanic_df[0] 같은 표현식은 오류를 발생한다. DataFrame 뒤의 []에는 칼럼명을 지정해야 하는데, 0이 칼럼 명이 아니기 때문이다. titanic_df[0, 0]이나 titanic_df[[0, 1, 2]]도 마찬가지 이유로 오류가 발생한다.

In [ ]:
print('단일 칼럼 데이터 추출:\n', titanic_df['Pclass'].head(3))
print('\n여러 칼럼의 데이터 추출:\n', titanic_df[['Survived', 'Pclass']].head(3))
# print('[] 안에 숫자 index는 KeyError 오류 발생:\n', titanic_df[0])

단일 칼럼 데이터 추출:
 0    3
1    1
2    3
Name: Pclass, dtype: int64

여러 칼럼의 데이터 추출:
    Survived  Pclass
0         0       3
1         1       1
2         1       3


앞에서 DataFrame의 [] 내에 숫자 값을 입력할 경우 오류가 발생한다고 했는데, 판다스의 인덱스 형태로 변환 가능한 표현식은 [] 내에 입력할 수 있다. 가령 titanic_df의 처음 2개 데이터를 추출하고자 titanic_df[0:2]와 같은 **슬라이싱**을 이용하면 정확히 원하는 결과를 반환한다.

In [ ]:
titanic_df[0:2]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0  A/5 21171   7.2500   NaN        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0   PC 17599  71.2833   C85        C

불린 인덱싱 표현도 가능하다. [] 내의 불린 인덱싱 기능은 원하는 데이터를 편리하게 추출해주므로 매우 자주 사용된다.
* Pclass 칼럼 없이 3인 데이터 3개만 추출한다.

In [ ]:
titanic_df[titanic_df['Pclass']==3].head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket   Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171  7.250   NaN        S
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...  7.925   NaN        S
4            5         0       3  Allen, Mr. ...    male  35.0      0      0          373450  8.050   NaN        S

* DataFrame 바로 뒤의 [] 연산자는 넘파이의 []나 Series의 []와 다르다.
* DataFrame 바로 뒤의 [] 내 입력 값은 칼럼명(또는 칼럼의 리스트)을 지정해 칼럼 지정 연산에 사용하거나 불린 인덱스 용도로만 사용해야 한다.
* DataFrame[0:2]와 같은 슬라이싱 연산으로 데이터를 추출하는 방법은 사용하지 않는 게 좋다.

### 명칭 기반 인덱싱과 위치 기반 인덱싱의 구분
명칭(Label) 기반 인덱싱은 칼럼의 명칭을 기반으로 위치를 지정하는 방식이다. '칼럼 명' 같이 명칭으로 열 위치를 지정하는 방식이다. 위치(Position) 기반 인덱싱은 0을 출발점으로 하는 가로축, 세로축 좌표 기반의 행과 열 위치를 기반으로 데이터를 지정한다. 따라서 행, 열 값으로 정수가 입력된다.

data_df를 reset_index()를 수행해 이번에는 기존 인덱스를 칼럼화하고 0부터 시작하는 새로운 인덱스를 추가하겠다. 그리고 이 새로운 인덱스를 다시 1부터 시작하는 인덱스로 바꿔보겠다.

In [ ]:
data={'Name':['Chulmin', 'Eunkyung', 'Jinwoong', 'Soobeom'],
      'Year':['Male', 'Female', 'Male', 'Male']
      }
data_df=pd.DataFrame(data, index=['one', 'two', 'three', 'four'])
data_df

Name    Year
one     Chulmin    Male
two    Eunkyung  Female
three  Jinwoong    Male
four    Soobeom    Male

In [ ]:
# data_df를 reset_index()로 새로운 숫자형 인덱스를 생성
data_df_reset=data_df.reset_index()
data_df_reset=data_df_reset.rename(columns={'index':'old_index'})

# 인덱스값에 1을 더해서 1부터 시작하는 새로운 인덱스값 생성
data_df_reset.index=data_df_reset.index+1
data_df_reset

old_index      Name    Year
1       one   Chulmin    Male
2       two  Eunkyung  Female
3     three  Jinwoong    Male
4      four   Soobeom    Male

일반적으로 행 위치에 적용되는 인덱스값과 위치 기반 인덱싱이 integer형일 때 코드 작성에 혼선을 초래할 우려가 있다. 그렇기 때문에 새롭게 명칭 기반 인덱싱 연산자인 loc[]와 위치 기반 인덱싱인 iloc[] 연산자를 도입해 이들의 구분을 명확히 하게 되었다.

### DataFrame iloc[] 연산자

iloc[]는 위치 기반 인덱싱만 허용하기 때문에 행과 열 값으로 integer 또는 integer형의 슬라이싱, 팬시 리스트 값을 입력해줘야 한다(불린 인덱싱은 조건을 기술하므로 이에 제약받지 않는다). 가령 위 예제의 DataFrame인 data_df의 첫 번째 행, 첫 번째 열의 데이터를 추출하고자 한다면 data_df.iloc[0, 0]과 같이 행과 열 위치에 위치 기반 인덱싱 값을 입력하면 된다.

In [ ]:
data_df.iloc[0, 0]

'Chulmin'

iloc[]에 위치 인덱싱이 아닌 명칭을 입력하면 오류가 발생한다.

In [ ]:
# data_df.iloc[0, 'Name']

마찬가지 이유로 문자열 인덱스를 행 위치에 입력해도 오류가 발생한다.

In [ ]:
# data_df.iloc['one', 0]

In [ ]:
data_df_reset.iloc[0, 1]

'Chulmin'

iloc[0, 1]은 명확히 위치 기반 인덱싱이므로 첫 번째 행(0)의 두 번째 열(1)의 위치에 있는 값을 성공적으로 반환한다.

iloc[]는 슬라이싱과 팬시 인덱싱은 제공하나 명확한 위치 기반 인덱싱이 사용되어야 하는 제약으로 인해 불린 인덱싱은 제공하지 않는다.

### DataFrame loc[] 연산자
loc[]는 명칭 기반으로 데이터를 추출한다. 따라서 행 위치에는 DataFrame의 index 값을, 열 위치에는 칼럼 명을 입력해 준다.

* 인덱스값이 'one'인 행의 칼럼 명이 'name'인 데이터를 추출해 보자.

In [ ]:
data_df.loc['one', 'Name']

'Chulmin'

index가 숫자 형일 수 있기 때문에 명칭 기반이라고 무조건 문자열을 입력한다는 선입견을 가져서는 안된다.

In [ ]:
data_df_reset.loc[1, 'Name']

'Chulmin'

loc[0, 'Name']을 data_df_reset에 호출하면 인덱스값이 0인 행이 없으므로 오류를 반환한다.

In [ ]:
data_df_reset
# data_df_reset.loc[0, 'Name']

old_index      Name    Year
1       one   Chulmin    Male
2       two  Eunkyung  Female
3     three  Jinwoong    Male
4      four   Soobeom    Male

loc[]에 슬라이싱 기호 ':'을 적용할 때 한 가지 유의할 점이 있다. 일반적으로 슬라이싱을 '시작 값:종료 값'과 같이 지정하면 시작 값 ~ 종료 값-1까지의 범위를 의미한다. 즉, 0:3이면 0부터 2까지 (0, 1, 2)를 의미한다. 그런데 loc[]에 슬라이싱 기호를 적용하면 종료 값-1이 아니라 종료 값까지 포함하는 것을 의미한다. 이는 명칭 기반 인덱싱의 특성 때문이다. 명칭은 숫자 형이 아닐 수 있기 때문에 -1을 할 수가 없다.

In [ ]:
print('위치 기반 iloc slicing\n', data_df.iloc[0:1, 0], '\n')
print('명칭 기반 loc slicing\n', data_df.loc['one': 'two', 'Name'])

위치 기반 iloc slicing
 one    Chulmin
Name: Name, dtype: object 

명칭 기반 loc slicing
 one     Chulmin
two    Eunkyung
Name: Name, dtype: object


loc[]을 적용한 경우에는 2개의 행을 반환했지만, 위치 기반 iloc[0:1]은 0번째 행 위치에 해당하는 1개의 행을 반환했다. 특히 DataFrame의 인덱스가 정수형이면 loc[]를 이용할 때 조심해야 한다. 가령 다음과 같이 loc[1:2, 'name']에서 1:2는 DataFrame의 인덱스값에 따른 명칭 기반 인덱싱이므로 인덱스가 1, 2인 2개의 데이터를 반환한다.

In [ ]:
print(data_df_reset.loc[1:2, 'Name'])

1     Chulmin
2    Eunkyung
Name: Name, dtype: object


1. 가장 중요한 것은 명칭 기반 인덱싱과 위치 기반 인덱싱의 차이를 이해하는 것이다. DataFrame의 인덱스가 숫자 형일 경우 행 위치에 오는 숫자는 위치 기반 인덱싱이 아니라 명칭 기반 인덱싱의 DataFrame 인덱스를 가리킨다.
2. iloc[]는 위치 기반 인덱싱만 가능하다. 따라서 행과 열 위치 값으로 정수형 값을 지정해 원하는 데이터를 반환한다.
3. loc[]는 명칭 기반 인덱싱만 가능하다. 따라서 행 위치에 DataFrame 인덱스가 오며, 열 위치에는 칼럼 명을 지정해 원하는 데이터를 반환한다.
5. 명칭 기반 인덱싱에서 슬라이싱을 '시작점:종료점'으로 지정할 때 시작점에서 종료점을 포함한 위치에 있는 데이터를 반환한다.

#### 불린 인덱싱
새롭게 타이타닉 데이터 세트를 DataFrame으로 로드하고, 승객 중 나이(Age)가 60세 이상인 데이터를 추출해 보겠다. [] 연산자 내에 불린 조건을 입력하면 불린 인덱싱으로 자동으로 결과를 찾아준다.

In [ ]:
titanic_df=pd.read_csv('/content/sample_data/titanic_train.csv')
titanic_boolean=titanic_df[titanic_df['Age']>60]
print(type(titanic_boolean))
titanic_boolean

<class 'pandas.core.frame.DataFrame'>


PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch       Ticket      Fare        Cabin Embarked
33            34         0       2  Wheadon, Mr...    male  66.0      0      0   C.A. 24579   10.5000          NaN        S
54            55         0       1  Ostby, Mr. ...    male  65.0      0      1       113509   61.9792          B30        C
96            97         0       1  Goldschmidt...    male  71.0      0      0     PC 17754   34.6542           A5        C
116          117         0       3  Connors, Mr...    male  70.5      0      0       370369    7.7500          NaN        Q
170          171         0       1  Van der hoe...    male  61.0      0      0       111240   33.5000          B19        S
252          253         0       1  Stead, Mr. ...    male  62.0      0      0       113514   26.5500          C87        S
275          276         1       1  Andrews, Mi...  female  63.0      1      0        13502   77.9583           D7        S
280          281         0       3  Duane, Mr. ...    male  65.0      0      0       336439    7.7500          NaN        Q
326          327         0       3  Nysveen, Mr...    male  61.0      0      0       345364    6.2375          NaN        S
438          439         0       1  Fortune, Mr...    male  64.0      1      4        19950  263.0000  C23 C25 C27        S
456          457         0       1  Millet, Mr....    male  65.0      0      0        13509   26.5500          E38        S
483          484         1       3  Turkula, Mr...  female  63.0      0      0         4134    9.5875          NaN        S
493          494         0       1  Artagaveyti...    male  71.0      0      0     PC 17609   49.5042          NaN        C
545          546         0       1  Nicholson, ...    male  64.0      0      0          693   26.0000          NaN        S
555          556         0       1  Wright, Mr....    male  62.0      0      0       113807   26.5500          NaN        S
570          571         1       2  Harris, Mr....    male  62.0      0      0  S.W./PP 752   10.5000          NaN        S
625          626         0       1  Sutton, Mr....    male  61.0      0      0        36963   32.3208          D50        S
630          631         1       1  Barkworth, ...    male  80.0      0      0        27042   30.0000          A23        S
672          673         0       2  Mitchell, M...    male  70.0      0      0   C.A. 24580   10.5000          NaN        S
745          746         0       1  Crosby, Cap...    male  70.0      1      1    WE/P 5735   71.0000          B22        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0       113572   80.0000          B28      NaN
851          852         0       3  Svensson, M...    male  74.0      0      0       347060    7.7750          NaN        S

반환된 titanic_boolean 객체의 타입은 DataFrame이다. [] 내에 불린 인덱싱을 적용하면 반환되는 객체가 DataFrame이므로 원하는 칼럼 명만 별도로 추출할 수 있다. 60세 이상인 승객의 나이와 이름만 추출해 보겠다. [] 연산자에 칼럼 명을 입력하면 된다. 칼럼이 두 개 이상이므로 [[]]를 사용한다.

In [ ]:
titanic_df[titanic_df['Age']>60][['Name', 'Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

loc[]을 이용해도 동일하게 적용할 수 있다. 단, ['Name', 'Age']는 칼럼 위치에 놓여야 한다.

In [ ]:
titanic_df.loc[titanic_df['Age']>60, ['Name', 'Age']].head(3)

Name   Age
33  Wheadon, Mr...  66.0
54  Ostby, Mr. ...  65.0
96  Goldschmidt...  71.0

여러 개의 복합 조건도 결합해 적용할 수 있다.
1. and 조건일 때는 &
2. or 조건일 때는 |
3. Not 조건일 때는 ~

* 나이가 60세 이상이고, 선실 등급이 1등급이며, 성별이 여성인 승객을 추출해 보겠다. 개별 조건은 ()로 묶고, 복합 조건 연산자를 사용하면 된다.

In [ ]:
titanic_df[(titanic_df['Age']>60)&(titanic_df['Pclass']==1)&(titanic_df['Sex']=='female')]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

개별 조건을 변수에 할당하고 이들 변수를 결합해서 불린 인덱싱을 수행할 수도 있다.

In [ ]:
cond1=titanic_df['Age']>60
cond2=titanic_df['Pclass']==1
cond3=titanic_df['Sex']=='female'
titanic_df[cond1&cond2&cond3]

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch  Ticket     Fare Cabin Embarked
275          276         1       1  Andrews, Mi...  female  63.0      1      0   13502  77.9583    D7        S
829          830         1       1  Stone, Mrs....  female  62.0      0      0  113572  80.0000   B28      NaN

이처럼 불린 인덱싱은 다재 다능하고 유연하다.

### 정렬, Aggregation 함수, GroupBy 적용

#### DataFrame, Series의 정렬 - sort_values()
DataFrame과 Series의 정렬을 위해서는 sort_values() 메서드를 이용한다. sort_values()는 RDBMS SQL의 order by 키워드와 매우 유사하다. sort_values()의 주요 입력 파라미터는 by, ascending, inplace이다. by로 특정 칼럼을 입력하면 해당 칼럼으로 정렬을 수행한다. ascending=True로 설정하면 오름차순으로 정렬하며, ascending=False로 설정하면 내림차순으로 정렬한다. 기본은 ascending=True이다. inplace 옵션도 쓸 수 있다.

* titanic_df를 Name 칼럼으로 오름차순 정렬해 보자.

In [ ]:
titanic_sorted=titanic_df.sort_values(by=['Name'])
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch     Ticket   Fare Cabin Embarked
845          846         0       3  Abbing, Mr....    male  42.0      0      0  C.A. 5547   7.55   NaN        S
746          747         0       3  Abbott, Mr....    male  16.0      1      1  C.A. 2673  20.25   NaN        S
279          280         1       3  Abbott, Mrs...  female  35.0      1      1  C.A. 2673  20.25   NaN        S

여러 개의 칼럼으로 정렬하려면 by에 리스트 형식으로 정렬하려는 칼럼을 입력하면 된다.
* Pclass와 Name을 내림차순으로 정렬해 보자.

In [ ]:
titanic_sorted=titanic_df.sort_values(by=['Pclass', 'Name'], ascending=False)
titanic_sorted.head(3)

PassengerId  Survived  Pclass            Name   Sex   Age  SibSp  Parch    Ticket  Fare Cabin Embarked
868          869         0       3  van Melkebe...  male   NaN      0      0    345777   9.5   NaN        S
153          154         0       3  van Billiar...  male  40.5      0      2  A/5. 851  14.5   NaN        S
282          283         0       3  de Pelsmaek...  male  16.0      0      0    345778   9.5   NaN        S

#### Aggregation 함수 적용
DataFrame에서 min(), max(), sum(), count()와 같은 aggregation 함수의 적용은 RDBMS SQL의 aggregation 함수 적용과 유사하다. 다만 DataFrame의 경우 DataFrame에서 바로 aggregation을 호출할 경우 모든 칼럼에 해당 aggregation을 적용한다는 차이가 있다.
* titanic_df에 count()를 적용하면 모든 칼럼에 count()를 반환한다.

In [ ]:
titanic_df.count()

PassengerId    891
Survived       891
Pclass         891
Name           891
Sex            891
Age            714
SibSp          891
Parch          891
Ticket         891
Fare           891
Cabin          204
Embarked       889
dtype: int64

특정 칼럼에 aggregation 함수를 적용하기 위해서는 DataFrame에 대상 칼럼들만 추출해 aggregation을 적용하면 된다.

In [ ]:
titanic_df[['Age', 'Fare']].mean()

Age     29.699118
Fare    32.204208
dtype: float64

#### groupby() 적용

DataFrame의 groupby()는 RDBMS SQL의 groupby 키워드와 유사하면서도 다른 면이 있다. SQL 판다스 모두 groupby를 분석 작업에 매우 많이 활용한다. DataFrame의 groupby() 사용 시 입력 파라미터 by에 칼럼을 입력하면 대상 칼럼으로 groupby된다. DataFrame에 groupby()를 호출하면 DataFrameGroupBy라는 또 다른 형태의 DataFrame을 반환한다.
* groupby(by='Pclass')를 호출하면 Pclass 칼럼 기준으로 GroupBy된 DataFrame Groupby 객체를 반환한다.

In [ ]:
titanic_groupby=titanic_df.groupby(by='Pclass')
print(type(titanic_groupby))

<class 'pandas.core.groupby.generic.DataFrameGroupBy'>


SQL의 groub by와 다르게, DataFrame에 groupby()를 호출해 반환된 결과에 aggregation 함수를 호출하면 groupby() 대상 칼럼을 제외한 모든 칼럼에 해당 aggregation 함수를 적용한다.

In [ ]:
titanic_groupby=titanic_df.groupby('Pclass').count()
titanic_groupby

PassengerId  Survived  Name  Sex  Age  SibSp  Parch  Ticket  Fare  Cabin  Embarked
Pclass                                                                                    
1               216       216   216  216  186    216    216     216   216    176       214
2               184       184   184  184  173    184    184     184   184     16       184
3               491       491   491  491  355    491    491     491   491     12       491

SQL의 경우 group by 적용 시 여러 개의 칼럼에 aggregation 함수를 호출하려면 대상 칼럼을 모두 Select 절에 나열해야 하는 것이 다르다. DataFrame의 groupby()에 특정 칼럼만 aggregation 함수를 적용하려면 groupby()로 반환된 DataFrameGroupBy 객체에 해당 칼럼을 필터링한 뒤 aggregation 함수를 적용한다.
* titanic_df.groupby('Pcalss')로 반환된 DataFrameGroupBy 객체에 [['PassengerId', 'Survived']]로 필터링해 PassengerId와 Survived 칼럼에만 count()를 수행한다.

In [ ]:
titanic_groupby=titanic_df.groupby('Pclass')[['PassengerId', 'Survived']].count()
titanic_groupby

PassengerId  Survived
Pclass                       
1               216       216
2               184       184
3               491       491

DataFrame groupby()와 SQL의 group by 키워드의 또 다른 차이는 SQL의 경우 서로 다른 aggregation 함수를 적용할 경우에는 Select 절에 나열하기만 하면 되지만, DataFrame groupby()의 경우 적용하려는 여러 개의 aggregation 함수명을 DataFrameGroupBy 객체의 agg() 내에 인자로 입력해서 사용한다는 점이다.

In [ ]:
titanic_df.groupby('Pclass')['Age'].agg([max, min])

max   min
Pclass            
1       80.0  0.92
2       70.0  0.67
3       74.0  0.42

서로 다른 aggregation 함수를 groupby에서 호출하려면 groupby()는 agg()를 이용해 이 같은 처리가 가능한데, agg() 내에 입력 값으로 딕셔너리 형태로 aggregation이 적용될 칼럼들과 aggregation 함수를 입력한다.

In [ ]:
agg_format={'Age':'max', 'SibSp':'sum', 'Fare':'mean'}
titanic_df.groupby('Pclass').agg(agg_format)

Age  SibSp       Fare
Pclass                        
1       80.0     90  84.154687
2       70.0     74  20.662183
3       74.0    302  13.675550

### 결손 데이터 처리하기

판다스는 결손 데이터(Missing Data)를 처리하는 편리한 API를 제공한다. 결손 데이터는 칼럼에 값이 없는, 즉 NULL인 경우를 의미하며, 이를 넘파이의 NaN으로 표시한다. 기본적으로 머신러닝 알고리즘은 이 NaN 값을 처리하지 않으므로 이 값을 다른 값으로 대체해야 한다. 또한 NaN 값은 평균, 총합 등의 함수 연산 시 제외가 된다. 특정 칼럼의 100개 데이터 중 10개가 NaN 값일 경우, 이 칼럼의 평균 값은 90개 데이터에 대한 평균이다. NaN 여부를 확인하는 API는 isna()이며, NaN 값을 다른 값으로 대체하는 API는 fillna()이다.

#### isna()로 결손 데이터 여부 확인

isna()는 데이터가 NaN인지 아닌지 알려준다. DataFrame에 isna()를 수행하면 모든 칼럼의 값이 NaN인지 아닌지를 True나 False로 알려준다.

In [ ]:
titanic_df.isna().head(3)

PassengerId  Survived  Pclass   Name    Sex    Age  SibSp  Parch  Ticket   Fare  Cabin  Embarked
0        False     False   False  False  False  False  False  False   False  False   True     False
1        False     False   False  False  False  False  False  False   False  False  False     False
2        False     False   False  False  False  False  False  False   False  False   True     False

결손 데이터의 개수는 isna() 결과에 sum() 함수를 추가해 구할 수 있다. sum()을 호출 시 True는 내부적으로 숫자 1로, False는 숫자 0으로 변환되므로 결손 데이터의 개수를 구할 수 있다.

In [ ]:
titanic_df.isna().sum()

PassengerId      0
Survived         0
Pclass           0
Name             0
Sex              0
Age            177
SibSp            0
Parch            0
Ticket           0
Fare             0
Cabin          687
Embarked         2
dtype: int64

#### fillna()로 결손 데이터 대체하기
fillna()를 이용하면 결손 데이터를 편리하게 다른 값으로 대체할 수 있다.
* 타이타닉 데이터 세트의 'Cabin' 칼럼의 NaN 값을 'C000'으로 대체해 보자.

In [ ]:
titanic_df['Cabin']=titanic_df['Cabin'].fillna('C000')
titanic_df.head(3)

PassengerId  Survived  Pclass            Name     Sex   Age  SibSp  Parch          Ticket     Fare Cabin Embarked
0            1         0       3  Braund, Mr....    male  22.0      1      0       A/5 21171   7.2500  C000        S
1            2         1       1  Cumings, Mr...  female  38.0      1      0        PC 17599  71.2833   C85        C
2            3         1       3  Heikkinen, ...  female  26.0      0      0  STON/O2. 31...   7.9250  C000        S

주의해야 할 점은 fillna()를 이용해 반환 값을 다시 받거나 inplace=True 파라미터를 fillna()에 추가해야 실제 데이터 세트 값이 변경된다는 점이다. inplace 옵션도 사용할 수 있다.
* 'Age' 칼럼의 NaN 값을 평균 나이로, 'Embarked' 칼럼의 NaN 값을 'S'로 대체해 모든 결손 데이터를 처리해 보자.

In [ ]:
titanic_df['Age']=titanic_df['Age'].fillna(titanic_df['Age'].mean())
titanic_df['Embarked']=titanic_df['Embarked'].fillna('S')
titanic_df.isna().sum()

PassengerId    0
Survived       0
Pclass         0
Name           0
Sex            0
Age            0
SibSp          0
Parch          0
Ticket         0
Fare           0
Cabin          0
Embarked       0
dtype: int64

### apply lambda 식으로 데이터 가공

판다스는 apply 함수에 lambda 식을 결합해 DataFrame이나 Series의 레코드별로 데이터를 가공하는 기능을 제공한다. 판다스의 경우 칼럼에 일괄적으로 데이터 가공을 하는 것이 속도 면에서 더 빠르나 복잡한 데이터 가공이 필요할 경우 어쩔 수 없이 apply lambda를 이용한다. 먼저 lambda 식에 익숙하지 않은 사람들을 위해 lambda 식에 대해 설명하겠다. lambda 식은 파이썬에서 함수형 프로그래밍(functional programming)을 지원하기 위해 만들었다.

* 입력값의 제곱 값을 구해서 반환하는 get_square(a)라는 함수가 있다고 가정하자.

In [ ]:
def get_square(a):
    return a**2
print('3의 제곱은:', get_square(3))

3의 제곱은: 9


위의 함수는 def get_square(a): 와 같이 함수명과 입력 인자를 먼저 선언하고 이후 함수 내에서 입력 인자를 가공한 뒤 결괏값을 return과 같은 문법으로 반환해야 한다. lambda는 이러한 함수의 선언과 함수 내의 처리를 한 줄의 식으로 쉽게 변환하는 식이다.

In [ ]:
lambda_square=lambda x : x**2
print('3의 제곱은:', lambda_square(3))

3의 제곱은: 9


lambda x : x**2에서 ':'로 입력 인자와 반환될 입력 인자의 계산식을 분리한다. ':'의 왼쪽에 있는 x는 입력 인자를 가리키며, 오른쪽은 입력 인자의 계산식이다. 오른쪽의 계산식은 결국 반환 값을 의미한다.

lambda 식을 이용할 때 여러 개의 값을 입력 인자로 사용해야 할 경우에는 보통 map() 함수를 결합해서 사용한다.

In [ ]:
a=[1, 2, 3]
squares=map(lambda x:x**2, a)
list(squares)

[1, 4, 9]

판다스 DataFrame의 lambda 식은 파이썬의 이러한 lambda 식을 그대로 적용한 것이다. 그럼 DataFrame의 apply에 lambda 식을 적용해 데이터를 가공해 보자. 먼저 'Name' 칼럼의 문자열 개수를 별도의 칼럼인 'Name_len'에 생성해 보자.

In [ ]:
titanic_df['Name_len']=titanic_df['Name'].apply(lambda x:len(x))
titanic_df[['Name', 'Name_len']].head(3)

Name  Name_len
0  Braund, Mr....        23
1  Cumings, Mr...        51
2  Heikkinen, ...        22

Lambda 식에서 if else 절을 사용해 조금 더 복잡한 가공을 해보자. 나이가 15세 미만이면 'Child', 그렇지 않으면 'Adult'로 구분하는 새로운 칼럼 'Child_Adult'를 apply lambda를 이용해 만들어 보자.

In [ ]:
titanic_df['Child_Adult']=titanic_df['Age'].apply(lambda x:'Child' if x <=15 else 'Adult')
titanic_df[['Age', 'Child_Adult']].head(8)

Age Child_Adult
0  22.000000       Adult
1  38.000000       Adult
2  26.000000       Adult
3  35.000000       Adult
4  35.000000       Adult
5  29.699118       Adult
6  54.000000       Adult
7   2.000000       Child

lambda 식은 if else를 지원하는데, **if 절의 경우 if 식보다 반환 값을 먼저 기술해야 한다.** 이는 lambda 식 ':' 기호의 오른편에 반환 값이 있어야 하기 때문이다. 따라서 lambda x : if x <=15 'Child' else 'Adult'가 아니라 lambda x : 'Child' if x <= 15 else 'Adult'이다. else의 경우는 else 식이 먼저 나오고 반환 값이 나중에 오면 된다.

또 한 가지는 if, else만 지원하고 if, else if, else와 같이 else if는 지원하지 않는다. else if를 이용하기 위해서는 else 절을 ()로 내포해 () 내에서 다시 if else를 적용해 사용한다.
* 나이가 15세 이하이면 Child, 15세~60세 사이는 Adult, 61세 이상은 Elderly로 분류하는 'Age_Cat' 칼럼을 만들어 보자.

In [ ]:
titanic_df['Age_cat']=titanic_df['Age'].apply(lambda x : 'Child' if x <= 15
                                              else ('Adult' if x <= 60
                                                    else 'Elderly'))
titanic_df['Age_cat'].value_counts()

Adult      786
Child       83
Elderly     22
Name: Age_cat, dtype: int64

그러나 else if가 많이 나와야 하는 경우나 switch case 문의 경우는 이렇게 else를 계속 내포해서 쓰기에는 부담스럽다. 이 경우에는 아예 별도의 함수를 만드는 게 더 나을 수 있다.
* 5살 이하는 Baby, 12살 이하는 Child, 18살 이하는 Teenage, 25살 이하는 Student, 35살 이하는 Young Adult, 60세 이하는 Adult, 그리고 그 이상은 Elderly로 분류해 보겠다.

In [ ]:
# 나이에 따라 세분화된 분류를 수행하는 함수 생성
def get_category(age):
    cat=''
    if age<=5:cat='Baby'
    elif age<=12:cat='Child'
    elif age<=18:cat='Teenager'
    elif age<=25:cat='Student'
    elif age<=35:cat='Young Adult'
    elif age<=60:cat='Adult'
    else:cat='Elderly'
    return cat

# lambda 식에 위에서 생성한 get_category() 함수를 반환값으로 지정
# get_category(X)는 입력값으로 'Age' 칼럼 값을 받아서 해당하는 cat 반환
titanic_df['Age_cat']=titanic_df['Age'].apply(lambda x:get_category(x))
titanic_df[['Age', 'Age_cat']].head()

Age      Age_cat
0  22.0      Student
1  38.0        Adult
2  26.0  Young Adult
3  35.0  Young Adult
4  35.0  Young Adult

## 05. 정리
넘파이와 판다스는 머신러닝 프로그램을 직접 만들면서 이해하고 모르는 부분들을 인터넷 검색 등으로 찾아 보는 게 학습을 위한 제일 빠른 방법이다.